File: IA_dovo_r/6 _Reinforcement Learning/Thompson Sampling/Thompson_Samplin.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jul 25 21:18:00 2020
Thompson sampling 

@author: David Andrade / @Bits2Matter
"""


import numpy as np 
import matplotlib.pyplot as plt
import pandas as pd 

# Read dataset 
dataset = pd.read_csv("Ads_CTR_Optimisation.csv")


# Thompson sampling algorithm 
import random
N = 10000
d = 10

number_rewards_1 = [0] * d
number_rewards_0 = [0] * d 
ads_selected = []
total_reward = 0
for n in range(0, N):
    max_random = 0
    ad = 0
    for i in range(0, d):
        random_betha = random.betavariate(number_rewards_1[i] + 1, number_rewards_0[i] + 1)
            
        if random_betha > max_random:
            max_random = random_betha
            ad = i 
            
    ads_selected.append(ad)
    reward = dataset.values[n, ad] 
    if reward == 1:
        number_rewards_1[ad] = number_rewards_1[ad] + 1 
    else:
        number_rewards_0[ad] = number_rewards_0[ad] + 1
    total_reward = total_reward + reward


# Data vis
plt.hist(ads_selected)
plt.title("Hist of ads selections")
plt.xlabel("Ads")
plt.ylabel(" # of times each ad was selected")
plt.show()

File: IA_dovo_r/6 _Reinforcement Learning/Thompson Sampling/Thompson_Samplin.R
# Created on 7/25/2020

# Thompson sampling 
# author: David Andrade / @Bits2Matter


dataset = read.csv("Ads_CTR_Optimisation.csv")


N = 10000
d = 10 
number_rewards_1 = integer(d)
number_rewards_0 = integer(d)
ads_selected = integer(0)
total_reward = 0

for(n in 1:N){
  max_random = 0
  ad = 0
  
  for (i in 1:d) {
    random_betha = rbeta(n = 1,
                         shape1 = number_rewards_1[i] + 1,
                         shape2 = number_rewards_0[i] + 1)
    
    if(random_betha > max_random){
      max_random = random_betha
      ad = i 
    }
    
  }
  ads_selected = append(ads_selected, ad)
  reward = dataset[n, ad]
  if (reward == 1){
    number_rewards_1[ad] = number_rewards_1[ad] + 1 
  }else{
    number_rewards_0[ad] = number_rewards_0[ad] + 1
  total_reward = total_reward + reward
  }
  
  total_reward = total_reward + reward 
}

# Data vis hist 
hist(ads_selected,
     col = 'blue',
     main = 'Histogram of ads selections',
     xlab = 'Ads',
     ylab = 'Number of times each ad was selected')



